src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+    return result.error;
+  }
+  auto &web_usb_devices = result.value;
+  // Iterate over the exposed devices.
+  uint8_t devices_num = web_usb_devices["length"].as<uint8_t>();
+  for (uint8_t i = 0; i < devices_num; i++) {
+    auto web_usb_device = web_usb_devices[i];
+    // …
+    *devs = discovered_devs_append(*devs, dev);
+  }
+  return LIBUSB_SUCCESS;
+}
+```
+
+Most of the backend code uses `val` and `promise_result` in a similar way as already shown above. There were few more interesting hacks I've had to do to convert properties exposed by WebUSB to those expected by libusb, both in this method and in others handling data transfer, but those implementation details are less important for the purposes of this article. Make sure to check the code and the comments if you're interested.
+
+## Porting event loops to the web
+
+One more piece of `libusb` that I do want to stop on is event handling. As described in the previous article, most APIs in system languages like C are synchronous, and event handling is no exception. It's usually implemented via an infinite loop that "polls" (tries to read data or blocks execution until some data is available) from a set of external I/O sources, and, when at least one of those responds, passes that as an event to the corresponding handler. Once the handler has finished its job, the control returns to the loop, and it pauses for another poll.
+
+There are a couple of problems with this approach on the web.
+
+First of all, WebUSB doesn't and cannot expose raw handles of the underlying devices, so polling those directly is not an option. Second, libusb uses `eventfd` and `pipe` APIs for other events as well as for handling transfers on operating systems without raw device handles, but `eventfd` is not currently supported in Emscripten, and `pipe`, while supported, [currently doesn't conform to the spec](https://github.com/emscripten-core/emscripten/issues/13214) and can't wait for events.
+
+Finally, the biggest problem is that the web has its own event loop. This global event loop is used for any external I/O operations (including `fetch`, timers, or, in this case, WebUSB), and it invokes event or `Promise` handlers whenever corresponding operation has finished. Executing another, nested, infinite event loop will block the browser's event loop from ever progressing, which means that not only will the UI become unresponsive, but also that the code will never get notifications for the very same I/O events it's waiting for. This usually results in a deadlock, and that's what happened when I tried to use libusb in a demo, too - the page froze.
+
+Like with other blocking I/O, to port such event loops to the web, developers need to find a way to run those loops without blocking the main thread. One way is to refactor the application to handle I/O events in a separate thread and pass the results back to the main one. The other is to use Asyncify to pause the loop and wait for events in a non-blocking fashion.
```

Comment:
```suggestion
Like with other blocking I/O, to port such event loops to the web, developers need to run those loops without blocking the main thread. One way is to refactor the application to handle I/O events in a separate thread and pass the results back to the main one. The other is to use Asyncify to pause the loop and wait for events in a non-blocking fashion.
```

Reply:
"find a way" seems important in this context, because it's one of those things that are easier said than done.

Reply:
Do you think you need to tell your readers or do you think they already know? If the former, you should make the point explicit. If the latter, you don't need the extra text. How you phrase it is up to you, but here's an example. (Emphasis is so you can see what I added. I'm not recommending bolding it.)

>developers need to run those loops without blocking the main thread, **even if it can sometimes be difficult**.

Reply:
I do need to tell the readers since it's one of the gotchas many people switching to the web are not aware of.